acceptance: fix a bug in TestDockerCLI and deflake some tests

knz opened the pull request:
Fixes #15532.

See https://github.com/cockroachdb/cockroach/issues/15532#issuecomment-298329050 for the backstory. I needed to add a check loop in `stop_server` until the server process really terminated (to ensure when the rocksdb files are closed), but that didn't work at first, leading me to discover that `kill -CONT` never failed, even on properly dead servers, as I had expected -- because those server processes were remaining in zombie state. This patch ensures that the top-level bash process cleans up dead servers.

knz committed:
acceptance: fix a bug in TestDockerCLI and deflake some tests

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15537)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/cli_test.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-Kj3FoIzQRUICdb43uqy:-Kj3FoIzQRUICdb43uqz:b-w853ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4f1bf820f296b0b9d0530c620b95230d12ba0f8/pkg/acceptance/cli_test.go#L75)):*
> ```Go
> 			// upon by the PID 1 process when they terminate, lest they
> 			// remain forever in the zombie state. Unfortunately, Expect
> 			// does not contain code to do this. Bash does.
> ```

Very interesting. How did you track that down?

---

*[pkg/cli/interactive_tests/common.tcl, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-Kj3EHjyKIKoAxBpBBJh:-Kj3EHjyKIKoAxBpBBJi:b-xxvaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4f1bf820f296b0b9d0530c620b95230d12ba0f8/pkg/cli/interactive_tests/common.tcl#L91)):*
> ```Tcl
>     system "$argv quit"
>     # If after 5 seconds the server hasn't shut down, trigger an error.
>     system "set -x; i=0; while test \$i -lt 5; do ps aux; wait `cat server_pid`; kill -CONT `cat server_pid` || exit 0; echo still waiting; sleep 1; i=`expr \$i + 1`; done; echo 'server still running?'; exit 1"
> ```

I'm not familiar with `wait`. What happens if the specified pid does not exit? I would naively expect `wait` to block forever.

Not important: I prefer to use `for i in $(seq 1 5); do ... done` to perform a fixed number of loop iterations.  Also, you're using bash, but seem to be restricting yourself to mostly plain shell style (e.g.  `expr \$i + 1` vs `\$[i + 1]`). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-:-Kj3FkE9AcMkH1m5SmOc:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/cli_test.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-Kj3FoIzQRUICdb43uqy:-Kj3I-i07d6aak1jYjtZ:bldicus) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4f1bf820f296b0b9d0530c620b95230d12ba0f8/pkg/acceptance/cli_test.go#L75)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Very interesting. How did you track that down?
</blockquote></details>

I was seeing in the log file that the server was being stopped completely but meanwhile `kill -CONT` was reporting the process still existed, which didn't make sense.
So I added `ps aux` in the wait loop and that's where I saw these zombie process in the `ps` output, and then I was "why is there no process to wait() them".
I knew that orphan processes should be waited upon by PID 1 (any other process doesn't have permission to call wait() on arbitrary processes that they didn't create themselves). `ps` had told me that `expect` was PID 1, so I dug into the expect code and expect indeed only calls wait() for the processes it has spawned itself via system/spawn, not the grand-children. Then I tried `bash` as PID 1 and that worked (as I knew it worked from pas experience booting linux systems with bash as init process).

---

*[pkg/cli/interactive_tests/common.tcl, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-Kj3EHjyKIKoAxBpBBJh:-Kj3I1jd9YvhDN5j2e_L:b-h5sgj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4f1bf820f296b0b9d0530c620b95230d12ba0f8/pkg/cli/interactive_tests/common.tcl#L91)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not familiar with `wait`. What happens if the specified pid does not exit? I would naively expect `wait` to block forever.

Not important: I prefer to use `for i in $(seq 1 5); do ... done` to perform a fixed number of loop iterations.  Also, you're using bash, but seem to be restricting yourself to mostly plain shell style (e.g.  `expr \$i + 1` vs `\$[i + 1]`). 
</blockquote></details>

The `wait` is a remnant from my investigation, and should not have been there. I removed it. (The top-level shell will do this for us now)
I am not using bash unfortunately: expect's `system` uses `sh` to invoke the shell, in which case `bash` "forgets" about its extensions (I'd have used `for((i=0;i<5;i++)); do...` otherwise). (Another thing: the `[ ... ]` syntax is nice to read but has special meaning in TCL, that's why I stuck to `test`). I changed the code to use the `seq` thing though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537)*
<!-- Sent from Reviewable.io -->

tamird commented:
The reparenting behaviour you describe here sounds a lot like what https://github.com/Yelp/dumb-init is meant to address. Should we make our containers' entry point dumb-init rather than catching this one-off? I could be missing something.

We already use dumb-init in cockroachdb/builder, so I think this would just require changing cockroachdb/postgres-test.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-:-Kj3KlfZdapeFzpdy6Bm:bltc8x0)*
<!-- Sent from Reviewable.io -->

knz commented:
Interesting, I didn't know about `dumb-init`. @tamird is `dumb-init` already populated in the test containers?

knz commented:
Oh never mind I forgot we had a separate repo for that. Looking into it now.

tamird commented:
cockroachdb/builder: [yes](https://github.com/cockroachdb/cockroach/blob/master/build/Dockerfile#L13)
cockroachdb/cockroach: [no](https://github.com/cockroachdb/cockroach/blob/master/build/deploy/Dockerfile)
cockroachdb/postgres-test: [no](https://github.com/cockroachdb/postgres-test/blob/master/Dockerfile)

knz commented:
Do we need one for the `cockroach` docker image?

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/cli/interactive_tests/common.tcl, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537#-Kj3EHjyKIKoAxBpBBJh:-Kj3MCTjk5PMFboc1br1:bn1wz0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4f1bf820f296b0b9d0530c620b95230d12ba0f8/pkg/cli/interactive_tests/common.tcl#L91)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The `wait` is a remnant from my investigation, and should not have been there. I removed it. (The top-level shell will do this for us now)
I am not using bash unfortunately: expect's `system` uses `sh` to invoke the shell, in which case `bash` "forgets" about its extensions (I'd have used `for((i=0;i<5;i++)); do...` otherwise). (Another thing: the `[ ... ]` syntax is nice to read but has special meaning in TCL, that's why I stuck to `test`). I changed the code to use the `seq` thing though.
</blockquote></details>

:thumbsup:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15537)*
<!-- Sent from Reviewable.io -->

tamird commented:
> Do we need one for the cockroach docker image?

I think we do - that image is used to run (some of) the acceptance tests in the release build. We wouldn't want that image to stay flaky, would we?

knz commented:
@tamird I do not understand, the tests that use the `cockroach` image will only find that binary inside the container, and that's the only command they will ever run. There is no zombie process / wait problem in that case. Only tests that have grand-children processes have issues.

knz commented:
Anyway I am following up on using `dumb-init` in a separate issue https://github.com/cockroachdb/cockroach/issues/15542.

knz commented:
Merging this for now, so as to avoid further flakiness on master. (all the flakiness observed recently were from TestDockerCLI)

a-robinson commented:
> > Do we need one for the cockroach docker image?

> I think we do - that image is used to run (some of) the acceptance tests in the release build. We wouldn't want that image to stay flaky, would we?

No, we shouldn't add it to our official image. It'd deviate too much from expected norms for provided images. We just should never encourage the use of `--background` in the official image.
